conn_js.go

```diff
@@ -0,0 +1,196 @@
+		c.Call("removeEventListener", "message", *c.messageHandler)
+		c.messageHandler.Release()
+	}
+	if c.closeHandler != nil {
+		c.Call("removeEventListener", "close", *c.closeHandler)
+		c.closeHandler.Release()
+	}
+	return nil
+}
+
+func (c *Conn) LocalAddr() net.Addr {
+	// TODO(albrow): is there some way to get the local address?
+	return NewAddr("0.0.0.0:0")
+}
+
+func (c *Conn) RemoteAddr() net.Addr {
+	rawURL := c.Get("url").String()
+	withoutPrefix := strings.TrimPrefix(rawURL, "ws://")
+	withoutSuffix := strings.TrimSuffix(withoutPrefix, "/")
+	return NewAddr(withoutSuffix)
+}
+
+func (c *Conn) SetDeadline(t time.Time) error {
+	// Not supported
+	return nil
+}
+
+func (c *Conn) SetReadDeadline(t time.Time) error {
+	// Not supported
+	return nil
+}
+
+func (c *Conn) SetWriteDeadline(t time.Time) error {
+	// Not supported
+	return nil
+}
+
+// NewConn creates a Conn given a regular js/wasm WebSocket Conn.
+func NewConn(raw js.Value) *Conn {
+	conn := &Conn{
+		Value:        raw,
+		incomingData: make(chan []byte),
+		closeSignal:  make(chan struct{}),
+	}
+	conn.setUpHandlers()
+	return conn
+}
+
+func (c *Conn) setUpHandlers() {
+	// c.mut.Lock()
+	// defer c.mut.Unlock()
+	if c.messageHandler != nil {
+		// Message handlers already created. Nothing to do.
+		return
+	}
+	messageHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
+			// TODO(albrow): Currently we assume data is of type Blob. Really we
+			// should check binaryType and then decode accordingly.
+			blob := args[0].Get("data")
+			text := readBlob(blob)
+			c.incomingData <- []byte(text)
+		}()
+		return nil
+	})
+	c.messageHandler = &messageHandler
+	c.Call("addEventListener", "message", messageHandler)
+
+	closeHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		close(c.closeSignal)
+		return nil
+	})
+	c.closeHandler = &closeHandler
+	c.Call("addEventListener", "close", closeHandler)
+}
+
+func (c *Conn) waitForOpen() error {
+	openSignal := make(chan struct{})
+	handler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		close(openSignal)
+		return nil
+	})
+	defer func() {
+		c.Call("removeEventListener", "open", handler)
+	}()
+	defer handler.Release()
+	c.Call("addEventListener", "open", handler)
+	<-openSignal
+	return nil
+}
+
+func readBlob(blob js.Value) string {
+	// const reader = new FileReader();
+	// reader.addEventListener('loadend', (e) => {
+	//   const text = e.srcElement.result;
+	//   console.log(text);
+	// });
+	// reader.readAsText(blb);
+	reader := js.Global().Get("FileReader").New()
+	textChan := make(chan string)
+	loadEndFunc := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
+			text := args[0].Get("srcElement").Get("result").String()
+			textChan <- text
+		}()
+		return nil
+	})
+	defer loadEndFunc.Release()
+	reader.Call("addEventListener", "loadend", loadEndFunc)
+	reader.Call("readAsText", blob)
+	return <-textChan
```

Comment:
Error handling? What if "loadend" never fires?

Reply:
If JavaScript throws, we can catch it with `defer` and `recover`. I can add that here like I did in a few other functions.

However, I'm not sure how to handle the case of "loadend" not firing. We can use a `select` and `time.After` to timeout after some amount of time if "loadend" doesn't fire, but what should we do next? Trying to keep the connection open could mean missing data or out-of-order reads. Maybe we should just close the connection if this happens?

Reply:
If loadend fires on error, we should be fine. Otherwise, we'll need to register an event listener for that error and select on it and loadend.